mypyc/codegen/emitclass.py

```diff
@@ -257,21 +257,40 @@ def emit_line() -> None:
 
     # XXX: there is no reason for the __weakref__ stuff to be mixed up with __dict__
     if cl.has_dict:
-        # __dict__ lives right after the struct and __weakref__ lives right after that
-        # TODO: They should get members in the struct instead of doing this nonsense.
-        weak_offset = f"{base_size} + sizeof(PyObject *)"
-        emitter.emit_lines(
-            f"PyMemberDef {members_name}[] = {{",
-            f'{{"__dict__", T_OBJECT_EX, {base_size}, 0, NULL}},',
-            f'{{"__weakref__", T_OBJECT_EX, {weak_offset}, 0, NULL}},',
-            "{0}",
-            "};",
-        )
+        # TODO: Use PY_VERSION_HEX and include patchlevel.h
+        if emitter.capi_version < (3, 11):
+            # __dict__ lives right after the struct and __weakref__ lives right after that
+            # TODO: They should get members in the struct instead of doing this nonsense.
+            weak_offset = f"{base_size} + sizeof(PyObject *)"
+            emitter.emit_lines(
+                f"PyMemberDef {members_name}[] = {{",
+                f'{{"__dict__", T_OBJECT_EX, {base_size}, 0, NULL}},',
+                f'{{"__weakref__", T_OBJECT_EX, {weak_offset}, 0, NULL}},',
+                "{0}",
+                "};",
+            )
+
+            fields["tp_members"] = members_name
+            fields["tp_basicsize"] = f"{base_size} + 2*sizeof(PyObject *)"
+            fields["tp_dictoffset"] = base_size
+            fields["tp_weaklistoffset"] = weak_offset
+
+            # In CPython >=3.11, __dict__ slot is automatically set before the type. So
+            # we need to let CPython manage it.
+        else:
+            # __weakref__ lives right after the struct
+            # TODO: __weakref__ should get members in the struct instead of doing this nonsense.
```

Comment:
```suggestion
        else:
            # In CPython >=3.11, __dict__ slot is automatically set before the type. So
            # we need to let CPython manage it.
            # __weakref__ lives right after the struct
            # TODO: __weakref__ should get members in the struct instead of doing this nonsense.
```